From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Fri, 3 Jan 2014 16:02:01 +0100
Subject: [PATCH] drm: bridge/dw_hdmi: use rx sense for plug detection if hpd
 is unreliable

Due to the voltage divider on the HPD line, the HDMI connector on
imx6q-sabrelite doesn't reliably detect connected DVI monitors.
This patch allows to use the RX_SENSE signals as a workaround when
enabled by a boolean device tree property 'hpd-unreliable'.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/gpu/drm/bridge/synopsys/dw-hdmi.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
index 1fb82690a9f3..0e95eaacb156 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
@@ -155,6 +155,7 @@ struct dw_hdmi {
 	struct clk *iahb_clk;
 	struct clk *cec_clk;
 	struct dw_hdmi_i2c *i2c;
+	bool hpd_unreliable;
 
 	struct hdmi_data_info hdmi_data;
 	const struct dw_hdmi_plat_data *plat_data;
@@ -1703,7 +1704,12 @@ static void dw_hdmi_phy_disable(struct dw_hdmi *hdmi, void *data)
 enum drm_connector_status dw_hdmi_phy_read_hpd(struct dw_hdmi *hdmi,
 					       void *data)
 {
-	return hdmi_readb(hdmi, HDMI_PHY_STAT0) & HDMI_PHY_HPD ?
+	u8 stat = hdmi_readb(hdmi, HDMI_PHY_STAT0);
+
+	/* If HPD is not reliable, use RX sense as fallback */
+	return ((stat & HDMI_PHY_HPD) ||
+		(hdmi->hpd_unreliable &&
+		 ((stat & HDMI_PHY_RX_SENSE) == HDMI_PHY_RX_SENSE))) ?
 		connector_status_connected : connector_status_disconnected;
 }
 EXPORT_SYMBOL_GPL(dw_hdmi_phy_read_hpd);
@@ -1713,7 +1719,7 @@ void dw_hdmi_phy_update_hpd(struct dw_hdmi *hdmi, void *data,
 {
 	u8 old_mask = hdmi->phy_mask;
 
-	if (force || disabled || !rxsense)
+	if (force || (disabled && !hdmi->hpd_unreliable) || !rxsense)
 		hdmi->phy_mask |= HDMI_PHY_RX_SENSE;
 	else
 		hdmi->phy_mask &= ~HDMI_PHY_RX_SENSE;
@@ -3074,7 +3080,7 @@ void dw_hdmi_setup_rx_sense(struct dw_hdmi *hdmi, bool hpd, bool rx_sense)
 		 * If the RX sense status indicates we're disconnected,
 		 * clear the software rxsense status.
 		 */
-		if (!rx_sense)
+		if (!rx_sense && !hdmi->hpd_unreliable)
 			hdmi->rxsense = false;
 
 		/*
@@ -3137,7 +3143,9 @@ static irqreturn_t dw_hdmi_irq(int irq, void *dev_id)
 			mutex_unlock(&hdmi->cec_notifier_mutex);
 		}
 
-		if (phy_stat & HDMI_PHY_HPD)
+		if ((phy_stat & HDMI_PHY_HPD) ||
+		    (hdmi->hpd_unreliable &&
+		     (phy_stat & HDMI_PHY_RX_SENSE) == HDMI_PHY_RX_SENSE))
 			status = connector_status_connected;
 
 		if (!(phy_stat & (HDMI_PHY_HPD | HDMI_PHY_RX_SENSE)))
@@ -3512,6 +3520,8 @@ struct dw_hdmi *dw_hdmi_probe(struct platform_device *pdev,
 		goto err_iahb;
 	}
 
+	hdmi->hpd_unreliable = of_property_read_bool(np, "hpd-unreliable");
+
 	ret = devm_request_threaded_irq(dev, irq, dw_hdmi_hardirq,
 					dw_hdmi_irq, IRQF_SHARED,
 					dev_name(dev), hdmi);
